feat(audio): use PyAV instead of ffmpeg

413x1nkp opened the pull request:
Replaces system calls to ffmpeg and ffmpeg python package usage with PyAV (`av`)

413x1nkp committed:
feat(audio): use PyAV instead of ffmpeg

replaced usage of ffmpeg in favor of PyAV (`av`)

413x1nkp committed:
Merge branch 'fumiama:dev' into dev

413x1nkp commented:
EXTREMELY dirty at the moment. Going to move `process_audio()` to `infer.lib.audio`.

413x1nkp committed:
refactor(audio): store all of the audio related functions in the `infer.lib.audio`

refactors previous commit to have singular functions for each task, all located in `infer.lib.audio`

413x1nkp commented:
A bit cleaner now, but would like to get rid of:
```py
def downsample_audio(input_path: str, output_path: str, format: str) -> None:
    if not os.path.exists(input_path): return
    
    input_container = av.open(input_path)
    output_container = av.open(output_path, 'w')

    # Create a stream in the output container
    input_stream = input_container.streams.audio[0]
    output_stream = output_container.add_stream(format)

    output_stream.bit_rate = 128_000 # 128kb/s (equivalent to -q:a 2)

    # Copy packets from the input file to the output file
    for packet in input_container.demux(input_stream):
        for frame in packet.decode():
            for out_packet in output_stream.encode(frame):
                output_container.mux(out_packet)

    for packet in output_stream.encode():
        output_container.mux(packet)
    
    # Close the containers
    input_container.close()
    output_container.close()

    try: # Remove the original file
        os.remove(input_path)
    except Exception as e:
        print(f"Failed to remove the original file: {e}")
```
In favor of:
```py
def resample_audio(input_path: str, output_path: str, format: str, sr: int, layout: str) -> None:
    if not os.path.exists(input_path): return
    
    input_container = av.open(input_path)
    output_container = av.open(output_path, 'w')

    # Create a stream in the output container
    input_stream = input_container.streams.audio[0]
    output_stream = output_container.add_stream(format, rate=sr, layout=layout)
    
    resampler = AudioResampler(format, layout, sr)

    # Copy packets from the input file to the output file
    for packet in input_container.demux(input_stream):
        for frame in packet.decode():
            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
            resampled = resampler.resample(frame)
            for out_packet in output_stream.encode(resampled):
                output_container.mux(out_packet)

    for packet in output_stream.encode():
        output_container.mux(packet)
    
    # Close the containers
    input_container.close()
    output_container.close()

    try: # Remove the original file
        os.remove(input_path)
    except Exception as e:
        print(f"Failed to remove the original file: {e}")
```
The only issue is that the samplerate has to be grabbed from the file itself in order to be put into the function, since in cases where downsample_audio() would be used, we don't care about the samplerate and are not planning to change it. If resample_audio() will be the primary function for any resampling, then it would also have to accept bit_rate param, which is kind of bad, since we'd have to grab samplerate from EVERY audio file.

Not sure what would be an ideal solution tbh

413x1nkp committed:
fix(audio): remove downsample_audio from mdxnet.py

it is no longer needed, since it's imported from infer.lib.audio

413x1nkp committed:
docs: remove every ffmpeg mention in the documentation to avoid confusion

413x1nkp committed:
chore(requirements): remove ffmpeg-python and ffmpy from all requirements

fumiama commented on infer/lib/audio.py:
```diff
@@ -39,20 +40,82 @@ def load_audio(file: str, sr: int) -> np.ndarray:
+        for frame in container.decode(audio=0):
+            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
+            resampled = resampler.resample(frame)
+            decoded_audio.append(resampled.to_ndarray())
```
Maybe slow to append into a list in a loop. Better to pre-allocate it.

413x1nkp replied:
Pre-allocated in 2574ce5

fumiama commented on infer/modules/uvr5/modules.py:
```diff
@@ -46,27 +47,23 @@ def uvr(model_name, inp_root, save_root_vocal, paths, save_root_ins, agg, format
-                    info["streams"][0]["channels"] == 2
-                    and info["streams"][0]["sample_rate"] == "44100"
-                ):
+                container = av.open(inp_path)
```
Consider wrapping this into a function.

413x1nkp replied:
Wrapped in 60d33c2

fumiama reviewed (changes requested):
Good job. Have you tried the uvr5 work status on your PC?

413x1nkp commented:
> Good job. Have you tried the uvr5 work status on your PC?

Well, the commit i'm on at the HEAD has no issues loading audio, the problem is with librosa though

Traceback:
```
Traceback (most recent call last):
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/uvr5/modules.py", line 69, in uvr
    pre_fun._path_audio_(
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/uvr5/vr.py", line 120, in _path_audio_
    wav_instrument = spec_utils.cmb_spectrogram_to_wave(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/lib/uvr5_pack/lib_v5/spec_utils.py", line 395, in cmb_spectrogram_to_wave
    spectrogram_to_wave(
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/lib/uvr5_pack/lib_v5/spec_utils.py", line 293, in spectrogram_to_wave
    wave_left = librosa.istft(spec_left, hop_length=hop_length)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/util/decorators.py", line 88, in inner_f
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/core/spectrum.py", line 394, in istft
    dtype = util.dtype_c2r(stft_matrix.dtype)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/util/decorators.py", line 88, in inner_f
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/util/utils.py", line 2185, in dtype_c2r
    np.dtype(complex): np.dtype(np.float).type,
                                ^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/numpy/__init__.py", line 324, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'float'.
`np.float` was a deprecated alias for the builtin `float`. To avoid this error in existing code, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'cfloat'?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/uvr5/modules.py", line 77, in uvr
    pre_fun._path_audio_(
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/uvr5/vr.py", line 120, in _path_audio_
    wav_instrument = spec_utils.cmb_spectrogram_to_wave(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/lib/uvr5_pack/lib_v5/spec_utils.py", line 395, in cmb_spectrogram_to_wave
    spectrogram_to_wave(
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/lib/uvr5_pack/lib_v5/spec_utils.py", line 293, in spectrogram_to_wave
    wave_left = librosa.istft(spec_left, hop_length=hop_length)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/util/decorators.py", line 88, in inner_f
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/core/spectrum.py", line 394, in istft
    dtype = util.dtype_c2r(stft_matrix.dtype)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/util/decorators.py", line 88, in inner_f
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/librosa/util/utils.py", line 2185, in dtype_c2r
    np.dtype(complex): np.dtype(np.float).type,
                                ^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/numpy/__init__.py", line 324, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'float'.
`np.float` was a deprecated alias for the builtin `float`. To avoid this error in existing code, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'cfloat'?

```

Doesn't seem to have anything to do with PyAV, must be an internal error with librosa
Updating librosa is not viable since `torchcrepe` has a hard dependency on librosa==0.9.1

413x1nkp committed:
fix(audio): fix loading for UVR

wrapped gathering of META info from the stream into a function

fixes loading for UVR

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp commented:
There appears to be issues in the new code for audio loading that is used in preprocessing
```py
start preprocess
/home/alex/Documents/rvc-trash//Fade away3.mp3	-> Traceback (most recent call last):
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/train/preprocess.py", line 86, in pipeline
    audio = signal.lfilter(self.bh, self.ah, audio)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/scipy/signal/_signaltools.py", line 2156, in lfilter
    return _sigtools._linear_filter(b, a, x, axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unsupported operand type(s) for *: 'float' and 'av.audio.frame.AudioFrame'

/home/alex/Documents/rvc-trash//Fade away2.mp3	-> Traceback (most recent call last):
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/train/preprocess.py", line 86, in pipeline
    audio = signal.lfilter(self.bh, self.ah, audio)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/scipy/signal/_signaltools.py", line 2156, in lfilter
    return _sigtools._linear_filter(b, a, x, axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unsupported operand type(s) for *: 'float' and 'av.audio.frame.AudioFrame'

/home/alex/Documents/rvc-trash//Fade away1.mp3	-> Traceback (most recent call last):
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/train/preprocess.py", line 86, in pipeline
    audio = signal.lfilter(self.bh, self.ah, audio)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/scipy/signal/_signaltools.py", line 2156, in lfilter
    return _sigtools._linear_filter(b, a, x, axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unsupported operand type(s) for *: 'float' and 'av.audio.frame.AudioFrame'

/home/alex/Documents/rvc-trash//Fade Away.mp3	-> Traceback (most recent call last):
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/train/preprocess.py", line 86, in pipeline
    audio = signal.lfilter(self.bh, self.ah, audio)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/.venv/lib/python3.12/site-packages/scipy/signal/_signaltools.py", line 2156, in lfilter
    return _sigtools._linear_filter(b, a, x, axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unsupported operand type(s) for *: 'float' and 'av.audio.frame.AudioFrame'

end preprocess
```
I'll work on this, but I'm not entirely sure what the error even means, lol

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
fix(audio): use np.frombuffer() instead of direct conversion of the resampled frames

this fixes traceback on preprocessing

fumiama commented:
> Updating librosa is not viable since torchcrepe has a hard dependency on librosa==0.9.1

maybe need to fork torchcrepe and modify it.

fumiama commented:
> TypeError: unsupported operand type(s) for *: 'float' and 'av.audio.frame.AudioFrame'

maybe the return value of load_audio is still 'av.audio.frame.AudioFrame'

413x1nkp committed:
feat(audio): pre-allocate decoded_audio array in the load_audio function

this should improve performance, even if just a little

413x1nkp commented:
> maybe need to fork torchcrepe and modify it.

Not sure, but for now the UVR isn't working.
Weren't you planning on getting rid of it anyway?

413x1nkp commented:
> maybe the return value of load_audio is still 'av.audio.frame.AudioFrame'

Yeah, that was it. Fixed it in 2574ce5.

413x1nkp commented:
I've read through some documentation on PyAV, turns out it also relies on ffmpeg, so the documentation and docker image changes should be reverted.

413x1nkp committed:
Revert "docs: remove every ffmpeg mention in the documentation to avoid confusion"

This reverts commit 1e05bbce03f98fdbc165d94488d4f029f1e260c8.

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #4 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp committed:
fix(requirements): revert removal of ffmpeg in unitest.yml and Dockerfile

fumiama commented:
> Weren't you planning on getting rid of it anyway?

Actually I want to remove it.

fumiama commented:
> I've read through some documentation on PyAV, turns out it also relies on ffmpeg

But the ffmpeg is in the whl, so no outer ffmpeg is needed.

fumiama commented on infer/lib/audio.py:
```diff
@@ -39,19 +40,102 @@ def load_audio(file: str, sr: int) -> np.ndarray:
+        resampler = AudioResampler(format="fltp", layout="mono", rate=sr)
+
+        # AV stores duration in nanoseconds
+        decoded_audio = (
```
Well, I mean pre-allocate a `numpy` array and not use `np.concatenate` at last.

413x1nkp replied:
Done in 953b2d6

fumiama commented on infer/lib/audio.py:
```diff
@@ -39,19 +40,102 @@ def load_audio(file: str, sr: int) -> np.ndarray:
+        for frame in container.decode(audio=0):
+            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
+            resampled = resampler.resample(frame)
+            decoded_audio.append(np.array(resampled))
```
So the append can be a direct copy to that array, like
```py
decoded_audio[offset:offset+this_frame_len]
```

413x1nkp replied:
> So the append can be a direct copy to that array, like
> 
> ```python
> decoded_audio[offset:offset+this_frame_len]
> ```

So... we need to cache the offset after each call for this, right?

fumiama replied:
Yes, just like the i = x, i++ in an iteration.

413x1nkp replied:
Implemented by 953b2d6

fumiama commented on infer/lib/audio.py:
```diff
@@ -39,19 +40,102 @@ def load_audio(file: str, sr: int) -> np.ndarray:
         raise RuntimeError(f"Failed to load audio: {e}")
 
-    return np.frombuffer(out, np.float32).flatten()
+    return np.frombuffer(audio, dtype=np.float32).flatten()
```
If the dtype of decoded_audio is already float32, now no need to call frombuffer and flatten, just return decoded_audio.

413x1nkp replied:
so, something like this?
```py
return np.frombuffer(audio, dtype=np.float32).flatten() if not isinstance(audio, np.float32) else audio
```
or this?
```py
return np.frombuffer(audio, dtype=np.float32).flatten() if not isinstance(audio, np.ndarray) else audio
```

fumiama replied:
No, you just alloc a numpy array with dtype=np.float32 as decoded_audio, so no need to check anymore.

413x1nkp replied:
> No, you just alloc a numpy array with dtype=np.float32 as decoded_audio, so no need to check anymore.

So... basically declare decoded_audio like so:
```py
decoded_audio = np.array([], dtype=np.float32)
```
?

fumiama replied:
You can use np.zeros to allocate an array. Try to google or ask GPT are some good choices. 👍

413x1nkp replied:
> You can use np.zeros to allocate an array. Try to google or ask GPT are some good choices. 👍

I'll come back to this tomorrow, I'm EXTREMELY sleep deprived at the moment :sweat_smile: 

```py
def load_audio(file: str, sr: int) -> np.ndarray:
    if not hasattr(load_audio, "offset"):
        load_audio.offset = 0 # Cache offset between function calls
    
    if not Path(file).exists():
        raise FileNotFoundError(f"File not found: {file}")

    try:
        container = av.open(file)
        resampler = AudioResampler(format="fltp", layout="mono", rate=sr)

        num_samples = int(((container.duration * sr / container.bit_rate) // 1_000_000) + 1)
        print(num_samples)
        decoded_audio = np.zeros(num_samples, dtype=np.float32)

        for frame in container.decode(audio=0):
            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
            resampled = np.array(resampler.resample(frame))

            frame_length = resampled.shape[0]
            decoded_audio[load_audio.offset:load_audio.offset+frame_length] = resampled
            load_audio.offset += frame_length

        container.close()
    except Exception as e:
        raise RuntimeError(f"Failed to load audio: {e}")

    return decoded_audio[:load_audio.offset].flatten()
```

```bash
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/lib/audio.py", line 63, in load_audio
    raise RuntimeError(f"Failed to load audio: {e}")
RuntimeError: Failed to load audio: float() argument must be a string or a real number, not 'av.audio.frame.AudioFrame'
```

fumiama replied:
> I'll come back to this tomorrow, I'm EXTREMELY sleep deprived at the moment

You can do these things anytime you like. No strict deadline in this project. Of course, you don't need to reply to my msgs immediately.

413x1nkp replied:
Assuming this was resolved with 953b2d6

413x1nkp commented:
> > I've read through some documentation on PyAV, turns out it also relies on ffmpeg
> 
> But the ffmpeg is in the whl, so no outer ffmpeg is needed.

Wait... what? It is??
I'll revert the revert of the commit then to make so dockerfile and unitest don't install ffmpeg

413x1nkp committed:
Revert "fix(requirements): revert removal of ffmpeg in unitest.yml and Dockerfile"

This reverts commit e28a0eebb23c89205f130d4727e0d3afa5a7141b.

fumiama commented:
> Wait... what? It is??

You can have a test about it.

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
feat(audio): pre-allocate numpy array to store the AudioFrame data in ndarray of dtype float32

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp commented:
I've googled for a bit, the NumPy stuff is extremely complicated
Didn't expect C++-like memory management in Python, but I'm glad that it is here :smile: 

```py
def load_audio(file: str, sr: int) -> np.ndarray:
    if not Path(file).exists():
        raise FileNotFoundError(f"File not found: {file}")

    try:
        container = av.open(file)
        resampler = AudioResampler(format="fltp", layout="mono", rate=sr)

        # Estimated maximum total number of samples to pre-allocate the array
        audio_duration_sec = container.duration / 1_000_000  # Internally the PyAV stores duration in microseconds
        estimated_total_samples = int(audio_duration_sec * container.streams.audio[0].rate)
        decoded_audio = np.zeros(estimated_total_samples + 1, dtype=np.float32)

        offset = 0
        for frame in container.decode(audio=0):
            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
            resampled_frames = resampler.resample(frame)
            for resampled_frame in resampled_frames:
                frame_data = np.array(resampled_frame.to_ndarray()).flatten()
                decoded_audio[offset:offset+len(frame_data)] = frame_data
                offset += len(frame_data)

        # Truncate the array to the actual size
        decoded_audio = decoded_audio[:offset]
    except Exception as e:
        raise RuntimeError(f"Failed to load audio: {e}")

    return decoded_audio
```
I may have made a mistake in the size computation, but other than that - it works!

413x1nkp committed:
Merge pull request #5 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp commented:
After some testing, here's the estimated pre-allocated size to the offset comparison of the resulting ndarray:
```
Estimated total samples:  2022912
Offset:  1684784

Estimated total samples:  2805120
Offset:  2335984

Estimated total samples:  2881152
Offset:  2399984

Estimated total samples:  5685120
Offset:  4735984
```
The resulting array is truncated by the actual size, `offset`
```py
decoded_audio = decoded_audio[:offset]
```

The oddity is that the estimation overshoots consistently for the factor of around 1.2
But it undershoots for audio files that are longer than 1 minute long, for some reason?
```
>>> 2022912 / 1684784
1.2006951632968974
>>> 2805120 / 2335984
1.20083014267221
>>> 2881152 / 2399984
1.200488003253355
>>> 5685120 / 4735984
1.2004094608427731
```

The error margin is relatively big, I'll try to work on minimizing the wasted resources on the pre-allocation of decoded_audio ndarray.

413x1nkp commented:
Turns out it's not as accurate as I expected. The estimation looks sufficient for small audio files (1 minute or less), but it's not enough for bigger files

The current implementation will work as intended only with short audio files, but will fail on longer ones.

I've looked into dynamic resizing, while it's slower than full pre-allocation, it works, unlike the method above.

```py
def load_audio(file: str, sr: int) -> np.ndarray:
    if not Path(file).exists():
        raise FileNotFoundError(f"File not found: {file}")

    try:
        container = av.open(file)
        resampler = AudioResampler(format="fltp", layout="mono", rate=sr)

        # Estimated maximum total number of samples to pre-allocate the array
        audio_duration_sec: float = (
            container.duration / 1_000_000
        )  # Internally the PyAV stores duration in microseconds
        estimated_total_samples = int(
            audio_duration_sec * container.streams.audio[0].rate
        )

        decoded_audio = np.zeros(estimated_total_samples + 1, dtype=np.float32)

        offset = 0
        for frame in container.decode(audio=0):
            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
            resampled_frames = resampler.resample(frame)
            for resampled_frame in resampled_frames:
                frame_data = np.array(resampled_frame.to_ndarray()).flatten()
                end_index = offset + len(frame_data)
                
                # Check if decoded_audio has enough space, resize if not
                if end_index > decoded_audio.shape[0]:
                    decoded_audio = np.resize(decoded_audio, end_index + 1)
                
                decoded_audio[offset : end_index] = frame_data
                offset += len(frame_data)
        # Truncate the array to the actual size
        decoded_audio = decoded_audio[:offset]
    except Exception as e:
        raise RuntimeError(f"Failed to load audio: {e}")

    return decoded_audio
```

Here's the data i collected:
| file | estimated size | decoded size (actual size) | duration (seconds) | Error (estimated / actual) |
| :---: | :-------------------: | :-----------------------------------: | :--------------------------: | :----------------------------------: |
| 2023-07-24_18-20-22wav.wav | 23424 | 58520 | 1.464 | 0.4002734107997266 |
| Fade away3.mp3 | 2022912 | 1684784 | 42.144 | 1.2006951632968974 |
| Fade away1.mp3 | 2805120 | 2335984 | 58.44 | 1.20083014267221 |
| Fade away2.mp3 | 2881152 | 2399984 | 60.024 | 1.200488003253355 | 
| Fade Away.mp3 | 5685120 | 4735984 | 118.44 | 1.2004094608427731 |
| mashup95BPM.wav | 9744000 | 8119984 | 203.0 | 1.2000023645366789 |
| output_vocals.wav | 102196 | 255450 | 6.38725 |  0.40006263456645136 |
| remake.flac | 8605384 | 7171138 | 179.278854 | 1.2000025658410143 |
| roomba.wav | 96782 | 241915 | 6.048875 | 0.40006613893309634 |
| output.wav | 2589576 | 6473900 | 161.8485 | 0.40000247146233336 |

The error should be equal to 1.0 in the best case scenario. It's okay to overshoot and have a value in between 1.0 and 1.1, but more than that is not wanted

I'm not entirely sure why the error is so inconsistent...
The error seems to be around ~0.4 on WAV specifically, for some reason?
On mp3 the error is consistently seems to be ~1.2
on FLAC it's around ~1.2 as well...

@fumiama Please let me know what you think about this. Could this be codec-specific?

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
Merge branch 'dev' into dev

fumiama commented:
> Didn't expect C++-like memory management in Python ...

That's true. The loop in py itself is not so efficient but I think it is an interesting optimization, too.

> The estimation looks sufficient for small audio files (1 minute or less), but it's not enough for bigger files

Since the audio is resampled to the target sample rate, maybe the estimation should come from sr but not the original container.streams.audio[0].rate.

```py
estimated_total_samples = int(
      audio_duration_sec * sr + 0.5
)

...

print(estimated_total_samples, offset)
# 32640 32623 in a short wav file, mono.
# 3688960 3688944 in a long mp3 file, dual.
```

413x1nkp commented:
> Since the audio is resampled to the target sample rate, maybe the estimation should come from sr but not the original container.streams.audio[0].rate.
> 
> ```python
> estimated_total_samples = int(
>       audio_duration_sec * sr + 0.5
> )
> 
> ...
> 
> print(estimated_total_samples, offset)
> # 32640 32623 in a short wav file, mono.
> # 3688960 3688944 in a long mp3 file, dual.
> ```

Oh my god, that was it!
Here i was trying to calculate complex formula that should be used in order to find the perfect size, but as it turns out i was just stupid since i didn't try multiplying by `sr` instead of by original audio's rate :rofl: 

Thank You VERY much for pointing this out, I'm not sure what kind of monstrous code I'd write if You didn't point at that

Should i leave resizing code that runs if the estimated size is somehow insufficient? (just in case, who knows what might happen)

```py
        for frame in container.decode(audio=0):
            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
            resampled_frames = resampler.resample(frame)
            for resampled_frame in resampled_frames:
                frame_data = np.array(resampled_frame.to_ndarray()).flatten()
                end_index = offset + len(frame_data)
                
                # Check if decoded_audio has enough space, and resize if necessary <---- Done just in case estimated size is not enough for array
                if end_index > decoded_audio.shape[0]:
                    decoded_audio = np.resize(decoded_audio, end_index + 1)
                
                decoded_audio[offset : end_index] = frame_data
                offset += len(frame_data)
```

413x1nkp committed:
fix(audio): fix the decoded_audio size estimation

in estimated_total_samples we multiply by `sr` instead of `container.streams.audio[0].rate` since we want to estimate size of the OUTPUT file, not the input one. - Added dynamic resizing, in case something goes wrong and the size of decoded_audio is estimated incorrectly

Fixed function `load_audio` when the input audio's samplerate does not match the desired samplerate (`sr`)

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #6 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp committed:
refactor(audio): remove `clean_path()` function as it serves no purpose anymore

413x1nkp committed:
docs: remove everything related to ffmpeg

this includes everything except for formats support specification in the training_tips docs, since it has nothing to do with what ffmpeg does/did but rather what audio formats are supported (all the ones that ffmpeg supports!)

413x1nkp committed:
docs: fix order of the steps in preparation in the READMEs

413x1nkp commented:
I've worked on some docs mentioning ffmpeg, they should now not have ffmpeg ever mentioned (unless it's relevant to audio formats that are acceptable) and also have the proper formatting.

I think that should be all for this PR, but if You think that the dynamic resizing that is done in the middle of the loop just in case should go - please let me know!

fumiama reviewed (approved):
> Should i leave resizing code that runs if the estimated size is somehow insufficient? 

I think it's OK. The if-in-loop in py is not a big problem.

Thanks for all the efforts you made.
